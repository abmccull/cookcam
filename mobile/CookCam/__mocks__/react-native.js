const React = require('react');
const RN = jest.requireActual('react-native');

// Mock PixelRatio for StyleSheet
const PixelRatio = {
  roundToNearestPixel: (value) => Math.round(value),
  get: () => 2,
  getFontScale: () => 1,
  getPixelSizeForLayoutSize: (size) => size * 2,
};

// Override modules that cause issues in tests
module.exports = {
  ...RN,
  PixelRatio,
  StyleSheet: {
    ...RN.StyleSheet,
    create: (styles) => styles,
    flatten: (style) => {
      if (!style) return {};
      if (Array.isArray(style)) {
        return Object.assign({}, ...style.filter(s => s));
      }
      return style;
    },
    absoluteFill: {
      position: 'absolute',
      top: 0,
      left: 0,
      right: 0,
      bottom: 0,
    },
    absoluteFillObject: {
      position: 'absolute',
      top: 0,
      left: 0,
      right: 0,
      bottom: 0,
    },
    hairlineWidth: 1,
  },
  Platform: {
    ...RN.Platform,
    OS: 'ios',
    select: (obj) => obj.ios || obj.default,
    Version: 14,
    isTV: false,
    isTesting: true,
  },
  Dimensions: {
    get: (dim) => {
      if (dim === 'window') {
        return { width: 375, height: 812, scale: 2, fontScale: 1 };
      }
      if (dim === 'screen') {
        return { width: 375, height: 812, scale: 2, fontScale: 1 };
      }
      return { width: 375, height: 812 };
    },
    addEventListener: jest.fn(() => ({ remove: jest.fn() })),
    removeEventListener: jest.fn(),
    set: jest.fn(),
  },
  Animated: {
    ...RN.Animated,
    timing: (value, config) => ({
      start: (callback) => {
        value.setValue(config.toValue);
        callback && callback({ finished: true });
      },
      stop: jest.fn(),
      reset: jest.fn(),
    }),
    spring: (value, config) => ({
      start: (callback) => {
        value.setValue(config.toValue);
        callback && callback({ finished: true });
      },
      stop: jest.fn(),
      reset: jest.fn(),
    }),
    Value: class {
      constructor(value) {
        this._value = value;
      }
      setValue(value) {
        this._value = value;
      }
      interpolate(config) {
        return this;
      }
      addListener(callback) {
        return '1';
      }
      removeListener(id) {}
      removeAllListeners() {}
      stopAnimation(callback) {
        callback && callback(this._value);
      }
      resetAnimation(callback) {
        callback && callback(this._value);
      }
      setOffset(offset) {}
      flattenOffset() {}
      extractOffset() {}
    },
    View: RN.View,
    Text: RN.Text,
    Image: RN.Image,
    ScrollView: RN.ScrollView,
    createAnimatedComponent: (Component) => Component,
    parallel: (animations, config) => ({
      start: (callback) => {
        animations.forEach(anim => anim.start && anim.start());
        callback && callback({ finished: true });
      },
      stop: jest.fn(),
      reset: jest.fn(),
    }),
    sequence: (animations) => ({
      start: (callback) => {
        animations.forEach(anim => anim.start && anim.start());
        callback && callback({ finished: true });
      },
      stop: jest.fn(),
      reset: jest.fn(),
    }),
    loop: (animation) => ({
      start: (callback) => {
        animation.start && animation.start();
        callback && callback({ finished: true });
      },
      stop: jest.fn(),
      reset: jest.fn(),
    }),
    event: () => jest.fn(),
    add: jest.fn(),
    subtract: jest.fn(),
    divide: jest.fn(),
    multiply: jest.fn(),
    modulo: jest.fn(),
    diffClamp: jest.fn(),
  },
  NativeModules: {
    ...RN.NativeModules,
    UIManager: {
      RCTView: {
        directEventTypes: {},
      },
    },
    PlatformConstants: {
      forceTouchAvailable: false,
    },
    RNCNetInfo: {
      getCurrentState: jest.fn(() => Promise.resolve({
        isConnected: true,
        isInternetReachable: true,
        type: 'wifi'
      })),
      addListener: jest.fn(),
      removeListeners: jest.fn(),
    },
    StatusBarManager: {
      HEIGHT: 20,
      getHeight: jest.fn(),
      setColor: jest.fn(),
      setStyle: jest.fn(),
      setHidden: jest.fn(),
      setNetworkActivityIndicatorVisible: jest.fn(),
      setBackgroundColor: jest.fn(),
      setTranslucent: jest.fn(),
    },
    SettingsManager: {
      settings: {},
      setValues: jest.fn(),
      deleteValues: jest.fn(),
    },
    DeviceInfo: {
      Dimensions: {
        window: { width: 375, height: 812, scale: 2, fontScale: 1 },
        screen: { width: 375, height: 812, scale: 2, fontScale: 1 },
      },
    },
  },
  TouchableOpacity: React.forwardRef((props, ref) =>
    React.createElement('TouchableOpacity', { ...props, ref })
  ),
  TouchableHighlight: React.forwardRef((props, ref) =>
    React.createElement('TouchableHighlight', { ...props, ref })
  ),
  TouchableWithoutFeedback: React.forwardRef((props, ref) =>
    React.createElement('TouchableWithoutFeedback', { ...props, ref })
  ),
  TouchableNativeFeedback: {
    SelectableBackground: jest.fn(),
    SelectableBackgroundBorderless: jest.fn(),
    Ripple: jest.fn(),
    canUseNativeForeground: jest.fn(),
  },
  View: React.forwardRef((props, ref) =>
    React.createElement('View', { ...props, ref })
  ),
  Text: React.forwardRef((props, ref) =>
    React.createElement('Text', { ...props, ref })
  ),
  Image: React.forwardRef((props, ref) =>
    React.createElement('Image', { ...props, ref })
  ),
  ScrollView: React.forwardRef((props, ref) =>
    React.createElement('ScrollView', { ...props, ref })
  ),
  TextInput: React.forwardRef((props, ref) =>
    React.createElement('TextInput', { ...props, ref })
  ),
  Modal: React.forwardRef((props, ref) =>
    React.createElement('Modal', { ...props, ref })
  ),
  FlatList: React.forwardRef((props, ref) =>
    React.createElement('FlatList', { ...props, ref })
  ),
  SectionList: React.forwardRef((props, ref) =>
    React.createElement('SectionList', { ...props, ref })
  ),
  VirtualizedList: React.forwardRef((props, ref) =>
    React.createElement('VirtualizedList', { ...props, ref })
  ),
  Alert: {
    alert: jest.fn(),
    prompt: jest.fn(),
  },
  Keyboard: {
    addListener: jest.fn(() => ({ remove: jest.fn() })),
    removeListener: jest.fn(),
    dismiss: jest.fn(),
    scheduleLayoutAnimation: jest.fn(),
  },
  Linking: {
    openURL: jest.fn(() => Promise.resolve()),
    canOpenURL: jest.fn(() => Promise.resolve(true)),
    getInitialURL: jest.fn(() => Promise.resolve(null)),
    addEventListener: jest.fn(() => ({ remove: jest.fn() })),
    removeEventListener: jest.fn(),
  },
  AppState: {
    currentState: 'active',
    addEventListener: jest.fn(() => ({ remove: jest.fn() })),
    removeEventListener: jest.fn(),
  },
  AccessibilityInfo: {
    isScreenReaderEnabled: jest.fn(() => Promise.resolve(false)),
    addEventListener: jest.fn(() => ({ remove: jest.fn() })),
    removeEventListener: jest.fn(),
    announceForAccessibility: jest.fn(),
    setAccessibilityFocus: jest.fn(),
  },
  findNodeHandle: jest.fn(() => 1),
  unstable_batchedUpdates: (callback) => callback(),
};