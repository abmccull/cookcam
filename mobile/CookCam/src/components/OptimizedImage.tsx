import React, { useState } from "react";
import { View, StyleSheet, ImageStyle, Image } from "react-native";
import { ChefHat } from "lucide-react-native";

// Type definitions to match the original FastImage interface
type Source = {
  uri: string;
  headers?: Record<string, string>;
};

type ResizeMode = "contain" | "cover" | "stretch" | "center";

interface OptimizedImageProps {
  source: Source | { uri: string };
  style: ImageStyle | ImageStyle[];
  resizeMode?: ResizeMode;
  placeholder?: React.ReactNode;
  fallbackColor?: string;
  _priority?: "low" | "normal" | "high";
  onLoadStart?: () => void;
  onLoad?: () => void;
  onError?: () => void;
}

const OptimizedImage: React.FC<OptimizedImageProps> = ({
  source,
  style,
  resizeMode = "cover",
  placeholder,
  fallbackColor = "#E5E5E7",
  _priority = "normal", // Note: priority is ignored for standard Image
  onLoadStart,
  onLoad,
  onError,
}) => {
  const [_isLoading, setIsLoading] = useState(true); // Note: not used in UI but tracks state
  const [hasError, setHasError] = useState(false);

  const handleLoadStart = () => {
    setIsLoading(true);
    setHasError(false);
    onLoadStart?.();
  };

  const handleLoad = () => {
    setIsLoading(false);
    setHasError(false);
    onLoad?.();
  };

  const handleError = () => {
    setIsLoading(false);
    setHasError(true);
    onError?.();
  };

  // Show fallback UI if error or no source
  if (hasError || !source || (typeof source === "object" && !source.uri)) {
    return (
      <View style={[style, styles.fallbackContainer]}>
        {placeholder || <ChefHat size={40} color={fallbackColor} />}
      </View>
    );
  }

  return (
    <Image
      source={
        typeof source === "object" && "uri" in source
          ? { uri: source.uri }
          : source
      }
      style={style}
      resizeMode={resizeMode}
      onLoadStart={handleLoadStart}
      onLoad={handleLoad}
      onError={handleError}
    />
  );
};

const styles = StyleSheet.create({
  fallbackContainer: {
    backgroundColor: "rgba(229, 229, 231, 0.1)",
    justifyContent: "center",
    alignItems: "center",
  },
});

export default OptimizedImage;
