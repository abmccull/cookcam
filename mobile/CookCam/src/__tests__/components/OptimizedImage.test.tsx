// Set up global mocks
global.PixelRatio = {
  roundToNearestPixel: (value) => Math.round(value),
  get: () => 2,
  getFontScale: () => 1,
  getPixelSizeForLayoutSize: (size) => size * 2,
};

// Mock React Native
jest.mock('react-native', () => {
  const actualRN = jest.requireActual('react-native');
  actualRN.PixelRatio = global.PixelRatio;
  return {
    ...actualRN,
    PixelRatio: global.PixelRatio,
    StyleSheet: {
      create: (styles) => styles,
      flatten: (style) => style,
    },
  };
});

// Mock environment config
jest.mock('../../config/env', () => ({
  __esModule: true,
  default: () => ({
    SUPABASE_URL: "https://test.supabase.co",
    SUPABASE_ANON_KEY: "test-anon-key",
    API_BASE_URL: "https://test-api.cookcam.com",
  }),
}));

// Mock icons
jest.mock('lucide-react-native', () => ({
  ChefHat: 'ChefHat',
}));

import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react-native';
import OptimizedImage from '../../components/OptimizedImage';

describe('OptimizedImage', () => {
  const defaultProps = {
    source: { uri: 'https://example.com/image.jpg' },
    style: { width: 100, height: 100 },
  };

  const mockOnLoadStart = jest.fn();
  const mockOnLoad = jest.fn();
  const mockOnError = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('Rendering', () => {
    it('should render correctly with all props', () => {
      const { toJSON } = render(
        <OptimizedImage
          source={defaultProps.source}
          style={defaultProps.style}
          resizeMode="cover"
          fallbackColor="#FF0000"
          _priority="high"
          onLoadStart={mockOnLoadStart}
          onLoad={mockOnLoad}
          onError={mockOnError}
        />
      );
      expect(toJSON()).toMatchSnapshot();
    });

    it('should render Image component with correct source', () => {
      const { UNSAFE_getByType } = render(<OptimizedImage {...defaultProps} />);
      
      const image = UNSAFE_getByType('Image');
      expect(image.props.source).toEqual(defaultProps.source);
    });

    it('should apply custom styles', () => {
      const customStyle = { width: 200, height: 200, borderRadius: 10 };
      const { UNSAFE_getByType } = render(
        <OptimizedImage {...defaultProps} style={customStyle} />
      );
      
      const image = UNSAFE_getByType('Image');
      expect(image.props.style).toEqual(customStyle);
    });

    it('should handle array of styles', () => {
      const styles = [
        { width: 200, height: 200 },
        { borderRadius: 10 },
        { opacity: 0.8 }
      ];
      const { UNSAFE_getByType } = render(
        <OptimizedImage {...defaultProps} style={styles} />
      );
      
      const image = UNSAFE_getByType('Image');
      expect(image.props.style).toEqual(styles);
    });

    it('should apply resize mode', () => {
      const { UNSAFE_getByType } = render(
        <OptimizedImage {...defaultProps} resizeMode="contain" />
      );
      
      const image = UNSAFE_getByType('Image');
      expect(image.props.resizeMode).toBe('contain');
    });

    it('should default to cover resize mode', () => {
      const { UNSAFE_getByType } = render(<OptimizedImage {...defaultProps} />);
      
      const image = UNSAFE_getByType('Image');
      expect(image.props.resizeMode).toBe('cover');
    });

    it('should handle all resize modes', () => {
      const resizeModes: Array<"contain" | "cover" | "stretch" | "center"> = 
        ["contain", "cover", "stretch", "center"];
      
      resizeModes.forEach(mode => {
        const { UNSAFE_getByType } = render(
          <OptimizedImage {...defaultProps} resizeMode={mode} />
        );
        
        const image = UNSAFE_getByType('Image');
        expect(image.props.resizeMode).toBe(mode);
      });
    });
  });

  describe('Source Handling', () => {
    it('should handle source with uri string', () => {
      const { UNSAFE_getByType } = render(
        <OptimizedImage 
          source={{ uri: 'https://example.com/test.jpg' }} 
          style={defaultProps.style} 
        />
      );
      
      const image = UNSAFE_getByType('Image');
      expect(image.props.source).toEqual({ uri: 'https://example.com/test.jpg' });
    });

    it('should handle source with headers', () => {
      const sourceWithHeaders = {
        uri: 'https://example.com/image.jpg',
        headers: {
          'Authorization': 'Bearer token',
          'X-Custom-Header': 'value'
        }
      };
      
      const { UNSAFE_getByType } = render(
        <OptimizedImage source={sourceWithHeaders} style={defaultProps.style} />
      );
      
      const image = UNSAFE_getByType('Image');
      expect(image.props.source.uri).toBe('https://example.com/image.jpg');
    });

    it('should handle empty uri', () => {
      const { UNSAFE_getByType } = render(
        <OptimizedImage source={{ uri: '' }} style={defaultProps.style} />
      );
      
      // Should show fallback
      const view = UNSAFE_getByType('View');
      expect(view).toBeTruthy();
      expect(screen.UNSAFE_getByType('ChefHat')).toBeTruthy();
    });

    it('should handle null source', () => {
      const { UNSAFE_getByType, UNSAFE_queryByType } = render(
        <OptimizedImage source={null as any} style={defaultProps.style} />
      );
      
      // Should show fallback
      expect(UNSAFE_queryByType('Image')).toBeFalsy();
      expect(UNSAFE_getByType('View')).toBeTruthy();
      expect(UNSAFE_getByType('ChefHat')).toBeTruthy();
    });

    it('should handle undefined source', () => {
      const { UNSAFE_getByType, UNSAFE_queryByType } = render(
        <OptimizedImage source={undefined as any} style={defaultProps.style} />
      );
      
      // Should show fallback
      expect(UNSAFE_queryByType('Image')).toBeFalsy();
      expect(UNSAFE_getByType('View')).toBeTruthy();
      expect(UNSAFE_getByType('ChefHat')).toBeTruthy();
    });

    it('should handle source without uri property', () => {
      const { UNSAFE_getByType, UNSAFE_queryByType } = render(
        <OptimizedImage source={{} as any} style={defaultProps.style} />
      );
      
      // Should show fallback
      expect(UNSAFE_queryByType('Image')).toBeFalsy();
      expect(UNSAFE_getByType('View')).toBeTruthy();
    });
  });

  describe('Loading Events', () => {
    it('should call onLoadStart when image starts loading', () => {
      const { UNSAFE_getByType } = render(
        <OptimizedImage {...defaultProps} onLoadStart={mockOnLoadStart} />
      );
      
      const image = UNSAFE_getByType('Image');
      image.props.onLoadStart();
      
      expect(mockOnLoadStart).toHaveBeenCalledTimes(1);
    });

    it('should call onLoad when image loads successfully', () => {
      const { UNSAFE_getByType } = render(
        <OptimizedImage {...defaultProps} onLoad={mockOnLoad} />
      );
      
      const image = UNSAFE_getByType('Image');
      image.props.onLoad();
      
      expect(mockOnLoad).toHaveBeenCalledTimes(1);
    });

    it('should call onError when image fails to load', () => {
      const { UNSAFE_getByType } = render(
        <OptimizedImage {...defaultProps} onError={mockOnError} />
      );
      
      const image = UNSAFE_getByType('Image');
      image.props.onError();
      
      expect(mockOnError).toHaveBeenCalledTimes(1);
    });

    it('should handle all events in sequence', () => {
      const { UNSAFE_getByType } = render(
        <OptimizedImage 
          {...defaultProps} 
          onLoadStart={mockOnLoadStart}
          onLoad={mockOnLoad}
          onError={mockOnError}
        />
      );
      
      const image = UNSAFE_getByType('Image');
      
      // Simulate loading sequence
      image.props.onLoadStart();
      image.props.onLoad();
      
      expect(mockOnLoadStart).toHaveBeenCalledTimes(1);
      expect(mockOnLoad).toHaveBeenCalledTimes(1);
      expect(mockOnError).not.toHaveBeenCalled();
    });

    it('should reset error state on successful load', () => {
      const { UNSAFE_getByType, rerender } = render(
        <OptimizedImage {...defaultProps} />
      );
      
      let image = UNSAFE_getByType('Image');
      
      // Simulate error
      image.props.onError();
      
      // Should show fallback
      rerender(<OptimizedImage {...defaultProps} />);
      
      // Check fallback is shown
      expect(screen.UNSAFE_getByType('ChefHat')).toBeTruthy();
    });
  });

  describe('Error Handling', () => {
    it('should show fallback on error', () => {
      const { UNSAFE_getByType } = render(
        <OptimizedImage {...defaultProps} />
      );
      
      const image = UNSAFE_getByType('Image');
      image.props.onError();
      
      // Force re-render to see fallback
      const { UNSAFE_getByType: getByTypeAfter } = render(
        <OptimizedImage {...defaultProps} />
      );
      
      // Start fresh and trigger error
      const newImage = getByTypeAfter('Image');
      newImage.props.onError();
    });

    it('should show custom placeholder on error', () => {
      const CustomPlaceholder = () => <div>Custom Error</div>;
      
      const { UNSAFE_getByType } = render(
        <OptimizedImage 
          {...defaultProps} 
          placeholder={<CustomPlaceholder />}
        />
      );
      
      const image = UNSAFE_getByType('Image');
      image.props.onError();
    });

    it('should use custom fallback color', () => {
      const { UNSAFE_getByType } = render(
        <OptimizedImage 
          source={{ uri: '' }} 
          style={defaultProps.style}
          fallbackColor="#FF0000"
        />
      );
      
      // Should show fallback with custom color
      const chefHat = UNSAFE_getByType('ChefHat');
      expect(chefHat.props.color).toBe('#FF0000');
    });

    it('should use default fallback color', () => {
      const { UNSAFE_getByType } = render(
        <OptimizedImage 
          source={{ uri: '' }} 
          style={defaultProps.style}
        />
      );
      
      // Should show fallback with default color
      const chefHat = UNSAFE_getByType('ChefHat');
      expect(chefHat.props.color).toBe('#E5E5E7');
    });
  });

  describe('Placeholder', () => {
    it('should show custom placeholder when provided and error occurs', () => {
      const CustomPlaceholder = <div testID="custom-placeholder">Loading...</div>;
      
      const { UNSAFE_getByType } = render(
        <OptimizedImage 
          source={{ uri: '' }}
          style={defaultProps.style}
          placeholder={CustomPlaceholder}
        />
      );
      
      // Should show custom placeholder instead of default ChefHat
      expect(screen.getByTestId('custom-placeholder')).toBeTruthy();
    });

    it('should show ChefHat icon when no placeholder provided', () => {
      const { UNSAFE_getByType } = render(
        <OptimizedImage 
          source={{ uri: '' }}
          style={defaultProps.style}
        />
      );
      
      expect(UNSAFE_getByType('ChefHat')).toBeTruthy();
    });

    it('should apply fallback container styles', () => {
      const { UNSAFE_getByType } = render(
        <OptimizedImage 
          source={{ uri: '' }}
          style={{ width: 200, height: 200 }}
        />
      );
      
      const view = UNSAFE_getByType('View');
      expect(view.props.style).toEqual(expect.arrayContaining([
        { width: 200, height: 200 },
        expect.objectContaining({
          backgroundColor: "rgba(229, 229, 231, 0.1)",
          justifyContent: "center",
          alignItems: "center",
        })
      ]));
    });
  });

  describe('Priority Handling', () => {
    it('should accept low priority', () => {
      const { UNSAFE_getByType } = render(
        <OptimizedImage {...defaultProps} _priority="low" />
      );
      
      const image = UNSAFE_getByType('Image');
      expect(image).toBeTruthy();
    });

    it('should accept normal priority', () => {
      const { UNSAFE_getByType } = render(
        <OptimizedImage {...defaultProps} _priority="normal" />
      );
      
      const image = UNSAFE_getByType('Image');
      expect(image).toBeTruthy();
    });

    it('should accept high priority', () => {
      const { UNSAFE_getByType } = render(
        <OptimizedImage {...defaultProps} _priority="high" />
      );
      
      const image = UNSAFE_getByType('Image');
      expect(image).toBeTruthy();
    });

    it('should default to normal priority', () => {
      const { toJSON } = render(<OptimizedImage {...defaultProps} />);
      // Component should render without error
      expect(toJSON()).toBeTruthy();
    });
  });

  describe('State Management', () => {
    it('should track loading state', () => {
      const { UNSAFE_getByType } = render(
        <OptimizedImage {...defaultProps} />
      );
      
      const image = UNSAFE_getByType('Image');
      
      // Start loading
      image.props.onLoadStart();
      
      // Complete loading
      image.props.onLoad();
      
      // Image should still be rendered
      expect(image).toBeTruthy();
    });

    it('should track error state', () => {
      const { UNSAFE_getByType, rerender } = render(
        <OptimizedImage {...defaultProps} />
      );
      
      const image = UNSAFE_getByType('Image');
      
      // Trigger error
      image.props.onError();
      
      // Re-render with same props
      rerender(<OptimizedImage {...defaultProps} />);
      
      // Should show fallback icon when error occurs
      expect(screen.UNSAFE_getByType('ChefHat')).toBeTruthy();
    });

    it('should reset states on new load', () => {
      const { UNSAFE_getByType } = render(
        <OptimizedImage {...defaultProps} />
      );
      
      const image = UNSAFE_getByType('Image');
      
      // Simulate error
      image.props.onError();
      
      // Start new load
      image.props.onLoadStart();
      
      // Complete load
      image.props.onLoad();
      
      // Should still render image
      expect(image).toBeTruthy();
    });
  });

  describe('Edge Cases', () => {
    it('should handle rapid state changes', () => {
      const { UNSAFE_getByType } = render(
        <OptimizedImage 
          {...defaultProps}
          onLoadStart={mockOnLoadStart}
          onLoad={mockOnLoad}
          onError={mockOnError}
        />
      );
      
      const image = UNSAFE_getByType('Image');
      
      // Rapid state changes
      image.props.onLoadStart();
      image.props.onError();
      image.props.onLoadStart();
      image.props.onLoad();
      image.props.onLoadStart();
      image.props.onError();
      
      expect(mockOnLoadStart).toHaveBeenCalledTimes(3);
      expect(mockOnLoad).toHaveBeenCalledTimes(1);
      expect(mockOnError).toHaveBeenCalledTimes(2);
    });

    it('should handle undefined callbacks', () => {
      const { UNSAFE_getByType } = render(
        <OptimizedImage {...defaultProps} />
      );
      
      const image = UNSAFE_getByType('Image');
      
      // Should not throw when callbacks are undefined
      expect(() => {
        image.props.onLoadStart();
        image.props.onLoad();
        image.props.onError();
      }).not.toThrow();
    });

    it('should handle changing source', () => {
      const { UNSAFE_getByType, rerender } = render(
        <OptimizedImage {...defaultProps} />
      );
      
      let image = UNSAFE_getByType('Image');
      expect(image.props.source.uri).toBe('https://example.com/image.jpg');
      
      // Change source
      rerender(
        <OptimizedImage 
          source={{ uri: 'https://example.com/new-image.jpg' }}
          style={defaultProps.style}
        />
      );
      
      image = UNSAFE_getByType('Image');
      expect(image.props.source.uri).toBe('https://example.com/new-image.jpg');
    });

    it('should handle changing styles', () => {
      const { UNSAFE_getByType, rerender } = render(
        <OptimizedImage {...defaultProps} />
      );
      
      let image = UNSAFE_getByType('Image');
      expect(image.props.style).toEqual({ width: 100, height: 100 });
      
      // Change styles
      const newStyle = { width: 200, height: 200, borderRadius: 20 };
      rerender(
        <OptimizedImage 
          source={defaultProps.source}
          style={newStyle}
        />
      );
      
      image = UNSAFE_getByType('Image');
      expect(image.props.style).toEqual(newStyle);
    });

    it('should handle empty style object', () => {
      const { UNSAFE_getByType } = render(
        <OptimizedImage source={defaultProps.source} style={{}} />
      );
      
      const image = UNSAFE_getByType('Image');
      expect(image.props.style).toEqual({});
    });

    it('should handle very long URIs', () => {
      const longUri = 'https://example.com/' + 'a'.repeat(1000) + '.jpg';
      const { UNSAFE_getByType } = render(
        <OptimizedImage 
          source={{ uri: longUri }}
          style={defaultProps.style}
        />
      );
      
      const image = UNSAFE_getByType('Image');
      expect(image.props.source.uri).toBe(longUri);
    });
  });

  describe('Performance', () => {
    it('should render efficiently with minimal re-renders', () => {
      const { rerender } = render(
        <OptimizedImage {...defaultProps} />
      );
      
      // Re-render with same props
      rerender(<OptimizedImage {...defaultProps} />);
      
      // Should not cause issues
      expect(screen.UNSAFE_getByType('Image')).toBeTruthy();
    });

    it('should handle priority prop without affecting render', () => {
      const priorities: Array<"low" | "normal" | "high"> = ["low", "normal", "high"];
      
      priorities.forEach(priority => {
        const { toJSON } = render(
          <OptimizedImage {...defaultProps} _priority={priority} />
        );
        expect(toJSON()).toBeTruthy();
      });
    });
  });
});