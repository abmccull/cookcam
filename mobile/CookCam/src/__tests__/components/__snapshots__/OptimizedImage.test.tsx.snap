// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptimizedImage Rendering should render correctly with all props 1`] = `
<Image
  onError={[Function]}
  onLoad={[Function]}
  onLoadStart={[Function]}
  ref={null}
  resizeMode="cover"
  source={
    {
      "uri": "https://example.com/image.jpg",
    }
  }
  style={
    {
      "height": 100,
      "width": 100,
    }
  }
/>
`;
